[SERV-556, SERV-557] Implement HTTP request handlers for addJob and listJobs

markmatney opened the pull request.

Commit:
Implement and add test for listJobs

DRickard commented on src/test/java/edu/ucla/library/prl/harvester/JobRequestsFT.java:
```diff
@@ -0,0 +1,128 @@
+     * @param aContext A test context
+     */
+    @Test
+    void testListBeforeAdd(final Vertx aVertx, final VertxTestContext aContext) {
```
Uh... don't you need a test to prove the handler actually retrieves jobs when the db is populated?

markmatney replied:
This is why I implemented all the Institution ops in #35; because I think verifying each method's behavior is most clearly done in context with all the other methods. It seems like a chicken-and-egg problem: it's a lot easier to see that listJobs, getJob, updateJob, and removeJob are doing the right thing if you can addJob. But since [the last PR was too big](https://github.com/UCLALibrary/prl-harvester/pull/35#issuecomment-1403831391), I figured it would be better to scope each PR to a single ticket. However I can add more stuff if this PR is too bare.

markmatney replied:
@ksclarke What do you think? Do you want these changesets scoped to each Jira ticket, or OK to implement addJob in this PR?

ksclarke replied:
Up to you. In reflecting on that last ticket, I may have been overreacting on the size of it. It was nowhere near the size of some of the large PRs I've submitted (which should have been broken up). I think my reaction was in part due to the lumping approach of the tests, which seemed like a paradigm shift rather than a large amount of code, really.

markmatney replied:
Ok, I'll go ahead and add it in then.

markmatney replied:
(It was mostly stashed, anyway.)

markmatney commented on src/main/java/edu/ucla/library/prl/harvester/handlers/ListJobsHandler.java:
```diff
@@ -0,0 +1,48 @@
+                    .putHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON.toString())
+                    .end(responseBody.encode());
+        }).onFailure(details -> {
+            response.setStatusCode(HttpStatus.SC_INTERNAL_SERVER_ERROR).end(details.getMessage());
```
In this case, I think 500 is fine for all errors, since the nano service method simply runs a database SELECT query that takes no parameters. And I think it makes sense to always consider that an internal server error from the perspective of a client.

Re: additional logging, I'm happy to add whatever you like.

ksclarke replied:
If it were me I'd just log the stacktrace in case that additional information proves useful in figuring out why it's not working. So, no value between knowing the database is down vs. it's receiving too many requests vs. bad credentials are being used, etc.?

markmatney replied:
For all practical purposes, too many requests is not within the realm of possibility here since this is backing an admin UI driven by a small number of users. And if the app is configured with incorrect credentials, the MainVerticle will not start, because instantiation of the schedule store service will fail, because instantiating the connection pool requires correct creds.

markmatney replied:
If the db is down, I would expect the error message to indicate some problem with the connection. So I think for all practical purposes, it covers the bases as-is.

markmatney commented on src/main/java/edu/ucla/library/prl/harvester/handlers/ListJobsHandler.java:
```diff
@@ -0,0 +1,48 @@
+     */
+    private final HarvestScheduleStoreService myHarvestScheduleStoreService;
+
+    /**
```
What additional information do you think is needed here?

ksclarke replied:
Just saying what this does for the generated javadocs (instead of having a blank value in them)... it doesn't have to be complicated: "Creates a new handler for job list requests." or something is fine.

markmatney replied:
It just begins to seem redundant, since creating new instances is what constructors do. But okay.

ksclarke replied:
Yes, a lot of times it's evident what something does (at least the author), but I look at it as producing user-readable documentation (Javadocs).

markmatney replied:
Since this is not a library, I think it makes more sense to evaluate the code comments on the basis of how well they assist developers on the project, rather than trying to "fill out" the HTML-ized javadoc. And actually, I don't understand why that is part of the criteria at all.

ksclarke replied:
I think javadocs assist the users (in this case developers) of the project, both new and old. You're lead here though, you're welcome to disregard my suggestion. :-)

Commit:
Print stack trace on handler failure

Commit:
Subclass ServiceException in order to distinguish the causing service

Since AddJobHandler will be calling HarvestJobSchedulerService (which
calls HarvestScheduleStoreService), it needs to be able to tell which
service caused an error so that it can return an appropriate HTTP
response.

Commit:
Return ServiceExceptions from job scheduler service

Commit:
Implement and add tests for addJob

Commit:
Fix casing of JSON keys

Don't provide lastSuccessfulRun timestamp with new test job data, since
new jobs haven't been run yet.

ksclarke commented on src/test/java/edu/ucla/library/prl/harvester/JobRequestsFT.java:
```diff
@@ -0,0 +1,204 @@
+        }).onSuccess(result -> aContext.completeNow()).onFailure(aContext::failNow);
+    }
+
+    /**
```
Even on tests, I'm pro-javadocs. Even if you skip the description of the test, putting the `@throws AddressException` without documentation of when/why this happens is just skirting the rules which are requiring that they be there. If that's the route you want to go, I'd suggest changing the checkstyle (or PMD?) rule that's requiring javadocs on the tests (that seems better to me than doing the bare minimum just to get the rules to pass).

markmatney replied:
For the record, so am I. The PMD and Checkstyle configurations that were inherited from your template project don't actually apply the rules to `src/test` (which I assumed was the intended behavior), which is why there are examples like [this](https://github.com/UCLALibrary/prl-harvester/blob/98575f759e883ff44ba4e9e4187b85f383c26aa8/src/test/java/edu/ucla/library/prl/harvester/utils/TestUtils.java#L80-L84) which doesn't even list the thrown exceptions, as well as in other projects and in other ways (undocumented member variables being common). This also isn't the first PR of mine on this project where I've left off descriptions of why particular exceptions are thrown by a test method. Just saying; I'm happy to follow the same rules in `src/test` as we do in `src/main`, but some consistency would help my understanding of expectations.

ksclarke replied:
Hmm, thanks for that [this](https://github.com/UCLALibrary/prl-harvester/blob/98575f759e883ff44ba4e9e4187b85f383c26aa8/src/test/java/edu/ucla/library/prl/harvester/utils/TestUtils.java#L80-L84)... that looks like a bug in the upstream rules. My take is every param, return value, throws, etc. should have documentation (that it's better to err on the side of too much/redundant documentation than not enough). I knew it was possible to not actually include text if you included the tag, but didn't know it was possible to skip the throws tag altogether. I'll create a bug ticket on that project, thanks. It's also my desire to keep improving the rules so that they require actual text after the tag, but that's obviously not something I've gotten to yet.

ksclarke replied:
As for expectations, there is no requirement that this project has to follow the upstream parent project's rules, as we've already overridden one naming pattern. That's up to you.
